File: 124_Binary_Tree_Maximum_Path_Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    maxSum = None 
    def maxPathSum(self, root: TreeNode) -> int:
        self.maxSum = -float('inf')
        self.maxGain(root)
        return self.maxSum
    
    def maxGain(self, root):
        if root == None:
            return 0 
        leftGain = max(self.maxGain(root.left), 0)
        rightGain = max(self.maxGain(root.right), 0) 
        summ = root.val + leftGain + rightGain 
        self.maxSum = max(self.maxSum, summ) 
        return root.val + max(leftGain, rightGain) 
    

File: 124_Binary_Tree_Maximum_Path_Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxSum; 
    public int maxPathSum(TreeNode root) {
        //depth first serach 
        maxSum = Integer.MIN_VALUE; 
        maxGain(root); 
        return maxSum; 
    }
    int maxGain(TreeNode root){
        if(root == null) return 0; 
    
    
    int leftGain = Math.max(maxGain(root.left), 0); 
    int rightGain = Math.max(maxGain(root.right), 0); 
    int sum = root.val + leftGain + rightGain; 
    maxSum = Math.max(maxSum, sum); 
    return root.val + Math.max(leftGain, rightGain); 
    }
}
